tensorflow/core/kernels/dynamic_stitch_op.cc

```diff
@@ -20,9 +20,119 @@ limitations under the License.
+    const Tensor& indices = indices_inputs[input_num];
+    auto indices_vec = indices.flat<int32>();
+    const Tensor& data = data_inputs[input_num];
+    auto data_flat =
+            data.shaped<T, 2>({indices_vec.dimension(0), slice_size});
+
+    if (DataTypeCanUseMemcpy(DataTypeToEnum<T>::v())) {
+      T* merged_base = &merged_flat(0, 0);
+      const T* data_base = &data_flat(0, 0);
+      const size_t slice_bytes = slice_size * sizeof(T);
+      for (int i = 0; i < indices_vec.size(); i++) {
+        int32 index = internal::SubtleMustCopy(indices_vec(i));
+        OP_REQUIRES(
+                ctx, FastBoundsCheck(index, first_dim_size),
+                errors::InvalidArgument("indices[", i, "] is out of range"));
+        memcpy(merged_base + index * slice_size, data_base + i * slice_size,
+               slice_bytes);
+      }
+    } else {
+      Eigen::DSizes<Eigen::DenseIndex, 2> sizes(1, slice_size);
+      for (int i = 0; i < indices_vec.size(); i++) {
+        // Copy slice data[i] to merged[indices[i]]
+        Eigen::DSizes<Eigen::DenseIndex, 2> data_indices(i, 0);
+        int32 index = internal::SubtleMustCopy(indices_vec(i));
+        OP_REQUIRES(
+                ctx, FastBoundsCheck(index, first_dim_size),
+                errors::InvalidArgument("indices[", i, "] is out of range"));
+        Eigen::DSizes<Eigen::DenseIndex, 2> merged_indices(index, 0);
+        merged_flat.slice(merged_indices, sizes) =
+                data_flat.slice(data_indices, sizes);
+      }
+    }
+  }
+}
+
+#ifdef GOOGLE_CUDA
+
+template <typename T>
+void DynamicStitchGPUImpl(const Eigen::GpuDevice& gpu_device,
+                          const int32 slice_size, const int32 first_dim_size,
+                          const CudaDeviceArrayStruct<int>& input_indices,
+                          const CudaDeviceArrayStruct<const T*>& input_ptrs,
+                          T* output);
+
+
+// because the collision requirements, we have to deal with
+// collion first before send data to gpu kernel.
+template <typename T>
+void DynamicStitchGPU(OpKernelContext* c,
+                        const int32 first_dim_size, const int32 data_elements_size,
+                        const OpInputList& indices_inputs, const OpInputList& data_inputs,
+                        Tensor* merged) {
+
+  const int slice_size = merged->flat_outer_dims<T>().dimension(1);
+  CudaDeviceArrayOnHost<int32> indices_flat(c, first_dim_size);
+  CudaDeviceArrayOnHost<const T*> data_flat(c, data_elements_size);
+  OP_REQUIRES_OK(c, indices_flat.Init());
+  OP_REQUIRES_OK(c, data_flat.Init());
+  // initialize the indices_flat (-1 represents missing indices)
+  for (int i = 0; i < first_dim_size; ++i) {
+    indices_flat.Set(i, -1);
+  }
+
+  //data_flat index
+  int32 idx = 0;
+  //sum of indices_inputs[i].NumElements() for compute indicies_flat value.
+  int32 base_size = 0;
+  for (int i = 0; i < indices_inputs.size(); ++i) {
```

Comment:
@ekelsen I don't think we can do it just with one loop. because we have to check the indices' collision before sending data to gpu, so we have to scan all of the indices on cpu.